src/metabase/lib/aggregation.cljc

```diff
@@ -2,10 +2,187 @@
   (:refer-clojure :exclude [count distinct max min])
   (:require
    [metabase.lib.common :as lib.common]
+   [metabase.lib.metadata.calculation :as lib.metadata.calculation]
    [metabase.lib.options :as lib.options]
+   [metabase.lib.schema :as lib.schema]
+   [metabase.lib.schema.aggregation :as lib.schema.aggregation]
+   [metabase.lib.schema.common :as lib.schema.common]
+   [metabase.lib.util :as lib.util]
+   [metabase.shared.util.i18n :as i18n]
    [metabase.util.malli :as mu])
   #?(:cljs (:require-macros [metabase.lib.aggregation])))
 
+(mu/defn resolve-aggregation :- ::lib.schema.aggregation/aggregation
+  "Resolve an aggregation with a specific `index`."
+  [query        :- ::lib.schema/query
+   stage-number :- :int
+   index        :- ::lib.schema.common/int-greater-than-or-equal-to-zero]
+  (let [{aggregations :aggregation} (lib.util/query-stage query stage-number)]
+    (when (<= (clojure.core/count aggregations) index)
+      (throw (ex-info (i18n/tru "No aggregation at index {0}" index)
+                      {:index        index
+                       :query        query
+                       :stage-number stage-number})))
+    (nth aggregations index)))
+
+(defmethod lib.metadata.calculation/describe-top-level-key :aggregation
+  [query stage-number _k]
+  (when-let [aggregations (not-empty (:aggregation (lib.util/query-stage query stage-number)))]
+    (lib.util/join-strings-with-conjunction
+     (i18n/tru "and")
+     (for [aggregation aggregations]
+       (lib.metadata.calculation/display-name query stage-number aggregation)))))
+
+(defmethod lib.metadata.calculation/metadata :aggregation
+  [query stage-number [_ag opts index, :as aggregation-ref]]
+  (let [aggregation (resolve-aggregation query stage-number index)]
+    (merge
+     (lib.metadata.calculation/metadata query stage-number aggregation)
+     {:field_ref aggregation-ref}
+     (when (:base-type opts)
+       {:base_type (:base-type opts)}))))
+
+;;; TODO -- merge this stuff into `defop` somehow.
+
+(defmethod lib.metadata.calculation/display-name-method :aggregation
```

Comment:
`display-name-method` here rhymes with `clojure.core/print-method`. This was previously called `display-name*` but I decided this was more meaningful and makes it clearer when you should use `display-name` (the wrapper function with error handling and type checking) vs `display-name-method`... `display-name-method` makes it clearer that you should implement `display-name-method` but call `display-name` than before.